tools/mkmodel/mkmodel.c

```diff
@@ -150,85 +221,171 @@ void indices_write(FILE *out, uint32_t type, void *data, uint32_t count)
+        w32(out, model->nodes[i].parent);
+        w32(out, model->nodes[i].num_children);
+        w32(out, data_ofs);
+        assert(ftell(out)-(offset+(i*NODE_SIZE)) == NODE_SIZE);
+        fseek(out, offset+data_ofs, SEEK_SET);
+        for(uint32_t j=0; j<model->nodes[i].num_children; j++) {
+            w32(out, model->nodes[i].children[j]);
+        }
+        data_ofs += 4*model->nodes[i].num_children;
+        if(model->nodes[i].name) {
+            uint32_t name_len = strlen(model->nodes[i].name)+1;
+            fseek(out, offset+name_ofs, SEEK_SET);
+            fwrite(model->nodes[i].name, strlen(model->nodes[i].name)+1, 1, out);
+            name_ofs += name_len;   
+        }
+    }
+    return offset+name_ofs;
+}
 
-    // Write primitives
-    for (size_t i = 0; i < model->num_meshes; i++)
-    {
-        offset_primitives[i] = ftell(out);
-        mesh_t *mesh = &model->meshes[i];
-        for (size_t j = 0; j < mesh->num_primitives; j++)
-        {
-            primitive_t *primitive = &mesh->primitives[j];
-            all_primitives[cur_primitive] = primitive;
-            int primitive_start = ftell(out);
+uint32_t model64_write_meshes(model64_data_t *model, uint32_t offset, FILE *out)
+{
+    uint32_t primitive_ofs = MESH_SIZE*model->num_meshes;
+    uint32_t data_ofs = primitive_ofs+(PRIMITIVE_SIZE*model64_get_primitive_total(model));
+    for(uint32_t i=0; i<model->num_meshes; i++) {
+        fseek(out, offset+(i*MESH_SIZE), SEEK_SET);
+        w32(out, model->meshes[i].num_primitives);
+        w32(out, primitive_ofs);
+        assert(ftell(out)-(offset+(i*MESH_SIZE)) == MESH_SIZE);
+        for(uint32_t j=0; j<model->meshes[i].num_primitives; j++) {
+            primitive_t *primitive = &model->meshes[i].primitives[j];
+            fseek(out, offset+primitive_ofs, SEEK_SET);
             w32(out, primitive->mode);
-            attribute_write(out, &primitive->position,  &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 0]);
-            attribute_write(out, &primitive->color,     &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 1]);
-            attribute_write(out, &primitive->texcoord,  &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 2]);
-            attribute_write(out, &primitive->normal,    &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 3]);
-            attribute_write(out, &primitive->mtx_index, &vertices_placeholders[cur_primitive*ATTRIBUTE_COUNT + 4]);
+            uint32_t color_data_ofs = data_ofs+attribute_get_data_size(&primitive->position);
+            uint32_t texcoord_data_ofs = color_data_ofs+attribute_get_data_size(&primitive->color);
+            uint32_t normal_data_ofs = texcoord_data_ofs+attribute_get_data_size(&primitive->texcoord);
+            uint32_t mtx_index_data_ofs = normal_data_ofs+attribute_get_data_size(&primitive->normal);
+            uint32_t index_ofs = mtx_index_data_ofs+attribute_get_data_size(&primitive->mtx_index);
```

Comment:
I get that the placeholders might be a bit cumbersome to use and I assume that's why you moved the code away from them. But I do think that they make the code easier to reason about. Now, whenever something about the file layout changes, you always have to make sure that the precomputed offsets are still correct, otherwise you'll get bugs that are hard to track. That problem does not exist with the placeholder technique.

We could try and come up with something that would make the placeholders less verbose maybe.

Reply:
I think that would involve something with a hash map that records a pointer, the pointer offset, and a list of references. We can use the stb data structures to help implement it.